### PR TITLE
Cleanup the RStudio Init Script

### PR DESCRIPTION
**Pull request by damacus**
This exposes functions to the user that are hidden in the current Bash.

- We no longer do an useradd, as we were trying to add and run the user as the wrong UID (1000 instead of the standard 1001)
- Fix lots of indentation errors in the functions
- Stop using && to guard sed's and echos, use if, else instead which is more readable
- Always run init_conda on start (we now call it from within the start function) so we make sure it's always in the correct state.

### CODE REVIEW

**Comment by andylightfoot** [comment]
I'd say we need more comprehensive tests for the R Studio image before we start refactoring it. Without these it would be easy to accidentally introduce a subtle change that could have an high impact of the stability of the platform (we saw this with the previous change to install nbstripout as part of the R Studio image which looked pretty harmless but proved to cause us a lot of headaches)

**Comment by damacus** [comment]
I pulled this out into it's own commit for clarification. But you're right. Before this gets deployed to users we need more use cases & tests. I have the on another branch (see #65 for some of the test), but I'm also going to pull that into it's own PR and work on that over the next few days.

The problem with testing this script, is it affects live running infrastructure differently to local/test environments, so no matter what integration tests we write for the build stage of it, we still won't cover the edge cases that users are experiencing.

I think the most pragmatic thing to do here is to build and deploy the container for one of us to test and optionally ask some users to upgrade to the new image and test it out for us.

**Comment by andylightfoot** [comment]
Yes I think you are right, we are going to have to have a set of tests (or scenarios to test) in addition to the integration tests.

e.g. a completely new user, new conda environment etc

This is probably something we can get input from everyone on.

I'm definitely not against the refactoring which is very much needed and looks good, it is just after the previous experience we had with the nbstripout release and the hard work we've put into making the platform more stable I'm just extra cautious about releasing changes to the R Studio image without the confidence of a more comprehensive set of tests.

**Comment by damacus** [comment]
Totally agreed. I'll put this on hold temporarily until we can get [this ticket prioritised ](https://trello.com/c/z4VEz6Hj/729-add-a-comprehensive-conda-test-suite)

**Comment by davidread** [comment]
Lots of nice refactoring here. 👍 

Is this a bugfix?:
> we were trying to add and run the user as the wrong UID (1000 instead of the standard 1001)

**Comment by damacus** [comment]
> Lots of nice refactoring here. 👍
> 
> Is this a bugfix?:
> 
> > we were trying to add and run the user as the wrong UID (1000 instead of the standard 1001)

Yeah, and I should probably call it out in tests. There doesn't appear to be a good reason to use UID 1001 instead of 1000 (which is already enabled in the container, and it's the rstudio user).

If we did use, user 1000, it would match the joyvan user in the jupyter images.

That all said, if there is a good reason as to why we're using `1001` we should all it out in tests.

File: start.sh [hunk 1]
```diff
@@ -1,115 +1,104 @@
 #!/usr/bin/env bash
 set -ex
 
-GROUP=staff
-USER_UID=1001
-export HOME="/home/$USER"
+USER="rstudio"
+USER_UID="1000"
```
Comment (damacus) [inline start.sh:5]: As per my discussion with @davidread with UIDs and GIDs, I realise we can't actually make this change right now. We'd have to do some more serious refactoring and re-init'ing of containers and chown'ing

File: start.sh [hunk 2]
```diff
@@ -1,115 +1,104 @@
...
 
 function init_user() {
-  if [ "$(id -u "$USER" 2>/dev/null)" != $USER_UID ]; then
-      useradd -g $GROUP -u $USER_UID -d "/home/$USER" "$USER" || true
-  fi
-
-  # this is only for running locally - in-cluster, home directories are
-  # provided via NAS
-  echo "${USER}:${USER}" | chpasswd
-  if [ ! -d "/home/$USER" ]; then
```
[the remark this answers is not in the document]
Comment (damacus) [reply start.sh:21]: Only if we move to using the included scripts. 

It looks like it might be taken straight fromn the user_init script actually

```bash
# File: /etc/cont-init.d/userconf

## Add a password to user
echo "$USER:$PASSWORD" | chpasswd
```

File: start.sh [hunk 3]
```diff
@@ -1,115 +1,104 @@
...
-    echo "no existing environment found"
-    conda create --use-index-cache --clone root -n rstudio --copy -y \
-    && chown -R "${USER}:${GROUP}" "/home/$USER/.conda/"
-    chmod -R 775 "/home/$USER/.conda/envs/rstudio/bin" \
-    && chmod -R 775 "/home/$USER/.conda/envs/rstudio/lib/R/bin"
+  source /opt/conda/etc/profile.d/conda.sh
+
+  if [[ $(conda env list | grep 'rstudio') != 0 ]]; then
+    echo "no existing Conda environment found, creating..."
+    fix_or_initialize_conda_home "${USER_HOME}"
```
[the remark this answers is not in the document]
Comment (damacus) [reply start.sh:34]: We should probably move `conda create --use-index-cache --clone root -n rstudio --copy -y` to `fix_or_initialize_conda_home` so the function does what it says
